services/inputhost/exthost.go

```diff
@@ -853,3 +870,58 @@ func (conn *extHost) getState() *admin.InputDestExtent {
 
 	return ext
 }
+
+// stopWritePump is used to stop the write pump if the extent is active or has
+// been prepped for drain and  mark the extHost as "Draining"
+func (conn *extHost) stopWritePump() {
+	if conn.state <= extHostDrainPrepped {
+		close(conn.closeChannel)
+		// wait for the write pump to drain for some timeout period
+		if ret := common.AwaitWaitGroup(&conn.waitWriteWG, defaultWGTimeout); !ret {
+			conn.logger.Fatalf("unable to stop write pump; wait group timeout")
+		}
+		conn.state = extHostDraining
+	}
+}
+
+// isDrained returns true if the following conditions are true:
+// 	(1) extHost state is "extHostDraining"
+//	(2) we have seen the reply for the last sent sequence number
+// if not, it returns false
+// Note: RLock() is sufficient here.
+func (conn *extHost) isDrained(replySeqNo int64) bool {
+	conn.lk.RLock()
+	if conn.state != extHostDraining {
+		conn.lk.RUnlock()
+		return false
+	}
+
+	ret := false
+	if atomic.LoadInt64(&conn.seqNo) == replySeqNo {
+		ret = true
+	}
+
+	conn.lk.RUnlock()
+	return ret
+}
+
+// prepDrain just decrements the number of writable extents for this destination
+func (conn *extHost) prepDrain() bool {
```

Comment:
Can we rename the prepDrain()/drain() and the associated states to `prepareForClose` / `closeWrite` ? drain is a bit confusing for extHost, because, it doesn't do anything special, it just prepares for close.